### ISSUE TITLE
Zero precision ,recall and f-score on custom dataset

### ISSUE THREAD

**Issue by acrosson** [marvis/pytorch-yolo2#21]
I trained for ~1600 epochs on a custom dataset with two classes. The losses have come down quite a bit, but I'm getting zeros for precision, recall and f-score.

What changes do I need to make to get this to train properly?

I have not updated the anchor boxes, can you explain how to do that? (how are they scaled)


```
1294386: nGT 21, recall 21, proposals 83, loss: x 0.191679, y 0.080800, w 0.062357, h 0.104740, conf 0.157618, cls 0.000881, total 0.598076                                                                 │·······
1294402: nGT 20, recall 19, proposals 97, loss: x 0.064738, y 0.085766, w 0.052172, h 0.031615, conf 0.412951, cls 0.001333, total 0.648575                                                                 │·······
1294418: nGT 24, recall 24, proposals 101, loss: x 0.120536, y 0.038962, w 0.029130, h 0.070185, conf 0.450639, cls 0.001438, total 0.710889                                                                │·······
1294434: nGT 21, recall 21, proposals 71, loss: x 0.039271, y 0.045842, w 0.052815, h 0.078578, conf 0.119643, cls 0.000321, total 0.336471                                                                 │·······
1294450: nGT 19, recall 19, proposals 120, loss: x 0.197666, y 0.194562, w 0.018538, h 0.019355, conf 0.376312, cls 0.003796, total 0.810228                                                                │·······
1294466: nGT 19, recall 19, proposals 100, loss: x 0.052779, y 0.051171, w 0.010564, h 0.032130, conf 0.054901, cls 0.005525, total 0.207069                                                                │·······
1294482: nGT 17, recall 17, proposals 100, loss: x 0.065116, y 0.045720, w 0.030157, h 0.038091, conf 0.568878, cls 0.002710, total 0.750672                                                                │·······
1294498: nGT 21, recall 21, proposals 91, loss: x 0.068207, y 0.049107, w 0.011280, h 0.035249, conf 0.052844, cls 0.000673, total 0.217361                                                                 │·······
1294514: nGT 24, recall 23, proposals 111, loss: x 0.086600, y 0.065052, w 0.088232, h 0.075294, conf 1.025374, cls 0.001391, total 1.341943                                                                │·······
1294530: nGT 19, recall 19, proposals 73, loss: x 0.026293, y 0.018219, w 0.017944, h 0.040603, conf 0.089713, cls 0.000803, total 0.193575                                                                 │·······
1294546: nGT 20, recall 20, proposals 80, loss: x 0.037674, y 0.056445, w 0.043170, h 0.043743, conf 0.152293, cls 0.000397, total 0.333722                                                                 │·······
1294562: nGT 21, recall 21, proposals 98, loss: x 0.055484, y 0.041236, w 0.023217, h 0.053546, conf 0.085894, cls 0.001428, total 0.260804                                                                 │·······
1294578: nGT 21, recall 20, proposals 66, loss: x 0.300329, y 0.064380, w 0.139514, h 0.908168, conf 0.327932, cls 0.967011, total 2.707335                                                                 │·······
1294594: nGT 27, recall 27, proposals 116, loss: x 0.060078, y 0.065306, w 0.030816, h 0.057140, conf 0.353824, cls 0.003778, total 0.570943                                                                │·······
1294610: nGT 22, recall 22, proposals 78, loss: x 0.031043, y 0.026628, w 0.026817, h 0.028361, conf 0.155648, cls 0.001030, total 0.269528                                                                 │·······
1294626: nGT 21, recall 21, proposals 102, loss: x 0.049257, y 0.034852, w 0.019491, h 0.016308, conf 0.079252, cls 0.001637, total 0.200797                                                                │·······
1294642: nGT 24, recall 24, proposals 131, loss: x 0.107006, y 0.199381, w 0.044377, h 0.073734, conf 0.803997, cls 0.002922, total 1.231417                                                                │·······
1294658: nGT 26, recall 26, proposals 104, loss: x 0.079384, y 0.030066, w 0.042223, h 0.064351, conf 0.229061, cls 0.001643, total 0.446728                                                                │·······
1294674: nGT 29, recall 28, proposals 93, loss: x 0.069832, y 0.050640, w 0.160470, h 0.633626, conf 0.355397, cls 0.008613, total 1.278579                                                                 │·······
1294690: nGT 21, recall 20, proposals 54, loss: x 0.047595, y 0.095901, w 0.606629, h 0.307124, conf 1.141145, cls 0.005670, total 2.204064                                                                 │·······
1294706: nGT 24, recall 24, proposals 78, loss: x 0.071578, y 0.041840, w 0.029242, h 0.062354, conf 0.321418, cls 0.001758, total 0.528190                                                                 │·······
1294722: nGT 28, recall 28, proposals 114, loss: x 0.107775, y 0.088004, w 0.038288, h 0.020913, conf 0.310151, cls 0.002962, total 0.568094                                                                │·······
1294738: nGT 23, recall 23, proposals 83, loss: x 0.078674, y 0.036628, w 0.022632, h 0.070169, conf 0.122063, cls 0.000869, total 0.331034                                                                 │·······
1294754: nGT 24, recall 24, proposals 82, loss: x 0.046268, y 0.055702, w 0.027008, h 0.031987, conf 0.112610, cls 0.001987, total 0.275563                                                                 │·······
1294770: nGT 25, recall 25, proposals 87, loss: x 0.044863, y 0.060390, w 0.067736, h 0.118288, conf 0.266852, cls 0.003266, total 0.561395                                                                 │·······
1294786: nGT 27, recall 25, proposals 91, loss: x 0.063308, y 0.174280, w 0.470813, h 0.080855, conf 1.178392, cls 0.298097, total 2.265744                                                                 │·······
1294802: nGT 20, recall 20, proposals 118, loss: x 0.239121, y 0.116060, w 0.080154, h 0.122954, conf 0.946752, cls 0.012245, total 1.517286                                                                │·······
1294818: nGT 20, recall 19, proposals 66, loss: x 0.219090, y 0.059595, w 0.097528, h 0.206593, conf 1.283862, cls 0.012689, total 1.879357                                                                 │·······
1294834: nGT 29, recall 28, proposals 133, loss: x 0.203282, y 0.216101, w 0.175833, h 0.070220, conf 1.545818, cls 0.007231, total 2.218486                                                                │·······
1294850: nGT 31, recall 31, proposals 134, loss: x 0.283963, y 0.500770, w 0.145383, h 0.152679, conf 0.729756, cls 0.041121, total 1.853672                                                                │·······
1294866: nGT 25, recall 24, proposals 95, loss: x 0.304713, y 0.062271, w 0.039418, h 0.193285, conf 0.847534, cls 0.012745, total 1.459966                                                                 │·······
1294882: nGT 29, recall 29, proposals 123, loss: x 0.061498, y 0.078497, w 0.091766, h 0.050161, conf 0.194632, cls 0.002413, total 0.478968                                                                │·······
1294898: nGT 23, recall 23, proposals 94, loss: x 0.061988, y 0.033381, w 0.056069, h 0.042701, conf 0.455985, cls 0.003712, total 0.653837                                                                 │·······
1294914: nGT 26, recall 26, proposals 95, loss: x 0.038289, y 0.066516, w 0.020397, h 0.061915, conf 0.329657, cls 0.000843, total 0.517617                                                                 │·······
1294930: nGT 21, recall 20, proposals 77, loss: x 0.043472, y 0.071030, w 0.023982, h 0.123312, conf 0.118349, cls 0.001612, total 0.381757                                                                 │·······
1294946: nGT 25, recall 25, proposals 100, loss: x 0.144314, y 0.194587, w 0.053056, h 0.066237, conf 0.315773, cls 0.001581, total 0.775547                                                                │·······
1294962: nGT 21, recall 21, proposals 121, loss: x 0.074221, y 0.101251, w 0.050594, h 0.053505, conf 0.215395, cls 0.001189, total 0.496154                                                                │·······
1294978: nGT 19, recall 19, proposals 132, loss: x 0.152955, y 0.194315, w 0.029937, h 0.061900, conf 0.572325, cls 0.006537, total 1.017969                                                                │·······
1294994: nGT 21, recall 21, proposals 117, loss: x 0.075395, y 0.076263, w 0.015889, h 0.041659, conf 0.331092, cls 0.003472, total 0.543770                                                                │·······
1295010: nGT 25, recall 25, proposals 128, loss: x 0.163195, y 0.172467, w 0.073641, h 0.159052, conf 1.594888, cls 0.068745, total 2.231988                                                                │·······
1295026: nGT 20, recall 19, proposals 77, loss: x 0.041927, y 0.059841, w 0.152994, h 0.106897, conf 0.201162, cls 0.000887, total 0.563707                                                                 │·······
1295042: nGT 24, recall 22, proposals 99, loss: x 0.117381, y 0.106450, w 0.264133, h 0.058498, conf 0.288284, cls 0.003960, total 0.838706                                                                 │·······
1295058: nGT 21, recall 20, proposals 62, loss: x 0.105285, y 0.111613, w 0.053934, h 0.140493, conf 0.206265, cls 0.808111, total 1.425701                                                                 │·······
1295074: nGT 23, recall 23, proposals 84, loss: x 0.026980, y 0.057341, w 0.023773, h 0.043864, conf 0.286903, cls 0.003113, total 0.441974                                                                 │·······
1295090: nGT 22, recall 22, proposals 96, loss: x 0.109033, y 0.082222, w 0.026835, h 0.025189, conf 0.105433, cls 0.011673, total 0.360384                                                                 │·······
1295106: nGT 19, recall 17, proposals 81, loss: x 0.031605, y 0.072685, w 0.136797, h 0.032506, conf 0.140581, cls 0.002034, total 0.416209                                                                 │·······
1295122: nGT 18, recall 18, proposals 93, loss: x 0.046847, y 0.033862, w 0.023247, h 0.043433, conf 0.334124, cls 0.002342, total 0.483856                                                                 │·······
1295138: nGT 23, recall 22, proposals 134, loss: x 0.165070, y 0.264870, w 1.806196, h 0.125330, conf 2.403664, cls 0.013194, total 4.778324                                                                │·······
1295140: nGT 3, recall 3, proposals 15, loss: x 0.007528, y 0.049220, w 0.000667, h 0.022014, conf 0.012180, cls 0.000103, total 0.091713                                                                   │·······
                                                                                                                                                                                                            │·······
2017-09-08 10:27:06 training with 50.599568 samples/s                                                                                                                                                       │·······
eval                                                                                                                                                                                                        │·······
0                                                                                                                                                                                                           │·······
[]                                                                                                                                                                                                          │·······
2017-09-08 10:27:06 precision: 0.000000, recall: 0.000000, fscore: 0.000000                                                                                                                                 │·······
2017-09-08 10:27:06 epoch 1682, processed 1295140 samples, lr 0.000010                                                                                                                                      │
```

**Comment by pgupta1204** [comment]
I am also getting the zero precision, recall and fscore while training on custom dataset? Have to figured out the issue?

**Comment by css1995** [comment]
you only need to change the  function  read_truths(lab_path)  in file utils.py
change `truths = truths.reshape(truths.size/5, 5)` to  `truths = truths.reshape(int(truths.size/5), 5)`

**Comment by edgarmedina1801** [comment]
I ran this line in evaluate mode and following the tutorial (python3).
run train.py cfg/voc.data cfg/yolo-voc.cfg darknet19_448.conv.23

but I got this result, is it fine? I thought this network was trained for a good accuracy. My configuration file in this (annexed), with no changes in the network. 
2018-03-09 14:20:40 precision: 0.000313, recall: 0.033416, fscore: 0.000620

[net]
# Testing
batch=16
subdivisions=8
# Training
# batch=64
# subdivisions=8
height=416
width=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.0001
burn_in=1000
max_batches = 80200
policy=steps
steps=-1,500,40000,60000
scales=0.1,10,.1,.1

**Comment by css1995** [comment]
sorry, I forgot how to solve this problem. But I remembered that it is the diferences between python 2 and python 3 that makes this wrong result. And there is an error in one .py file you need to modify, but sorry I fogot which one.






On 3/10/2018 01:18，edgarmedina1801<notifications@github.com> wrote：

I ran this line in evaluate mode and following the tutorial (python3).
run train.py cfg/voc.data cfg/yolo-voc.cfg darknet19_448.conv.23

but I got this result, is it fine? I thought this network was trained for a good accuracy. My configuration file in this (annexed), with no changes in the network.
2018-03-09 14:20:40 precision: 0.000313, recall: 0.033416, fscore: 0.000620

[net]

Testing

batch=16
subdivisions=8

Training
batch=64
subdivisions=8

height=416
width=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.0001
burn_in=1000
max_batches = 80200
policy=steps
steps=-1,500,40000,60000
scales=0.1,10,.1,.1

**Comment by sonack** [comment]
I think it is because your custom dataset's labpath changed, which is different from the VOC dataset organization.
In `image.py` and `dataset.py`. there is a line  `labpath = imgpath.replace('images', 'labels').replace('JPEGImages', 'labels').replace('.jpg', '.txt').replace('.png','.txt')
`, which may not work properly for your custom dataset.

I changed them to 
```
img_id = os.path.splitext(os.path.basename(imgpath))[0]
labpath = os.path.join(LABEL_PATH, img_id + ".txt")
```

 where the LABEL_PATH is your root directory storing all the generated label txt files. And all seems to work properly:

```
2018-05-22 14:49:27 precision: 0.642657, recall: 0.607427, fscore: 0.624540
```

**Comment by elixir-1** [comment]
@sonack Did you use custom dataset? 
I tried training on custom dataset and made the corresponding changes in code including labpath in image.py. In the training process, the proposals decrease to zero while recall is non-zero. Also, precision, recall and fscore remain zero. I have trained the model for only 3 epochs.

**Comment by sonack** [comment]
@elixir-1  Yes. I trained on my own custom dataset KITTI. I can not know what happened to you according to your description, but you can check if your `dataset.py` has also been updated?

**Comment by elixir-1** [comment]
@sonack, 
1. Which optimizer, learning rate did you use? I switched to Adam and now I get non zero fscore. I am using Adam with learning rate = 0.001 and scale = 0.25 for first 500 steps.

2. Do you understand the data augmentation part in image.py? Are they augmenting the original image and returning that?

3. Did you use the resizing strategy as it is or did you change it?

4. Why does the optimizer have learning rate and decay changed according to the batch size?

Thank you.
